docs/framework/interop/default-marshaling-for-strings.md

```diff
@@ -1,264 +1,280 @@
-|`UnmanagedType.LPTStr`|A pointer to a null-terminated array of platform-dependent characters.|  
-|`UnmanagedType.LPWStr`|A pointer to a null-terminated array of Unicode characters.|  
-|`UnmanagedType.TBStr`|A COM-style `BSTR` with a prefixed length and platform-dependent characters.|  
-|`VBByRefStr`|A value that enables Visual Basic .NET to change a string in unmanaged code, and have the results reflected in managed code. This value is supported only for platform invoke. This is default value in Visual Basic for `ByVal` strings.|  
-  
- This table applies to strings. However, for <xref:System.Text.StringBuilder>, the only options allowed are `LPStr`, `LPTStr`, and `LPWStr`.  
-  
- The following type definition shows the correct use of `MarshalAsAttribute` for platform invoke calls.  
-  
-```vb  
-Class StringLibAPI      
-Public Declare Auto Sub PassLPStr Lib "StringLib.Dll" _  
-(<MarshalAs(UnmanagedType.LPStr)> s As String)      
-Public Declare Auto Sub PassLPWStr Lib "StringLib.Dll" _  
-(<MarshalAs(UnmanagedType.LPWStr)> s As String)      
-Public Declare Auto Sub PassLPTStr Lib "StringLib.Dll" _  
-(<MarshalAs(UnmanagedType.LPTStr)> s As String)      
-Public Declare Auto Sub PassBStr Lib "StringLib.Dll" _  
-(<MarshalAs(UnmanagedType.BStr)> s As String)      
-Public Declare Auto Sub PassAnsiBStr Lib "StringLib.Dll" _  
-(<MarshalAs(UnmanagedType.AnsiBStr)> s As String)      
-Public Declare Auto Sub PassTBStr Lib "StringLib.Dll" _  
-(<MarshalAs(UnmanagedType.TBStr)> s As String)  
-End Class  
+## Fixed-Length String Buffers
+
+In some circumstances, a fixed-length character buffer must be passed into unmanaged code to be manipulated. Simply passing a string does not work in this case because the callee cannot modify the contents of the passed buffer. Even if the string is passed by reference, there is no way to initialize the buffer to a given size.
+
+The solution is to pass a <xref:System.Text.StringBuilder> buffer as the argument instead of a <xref:System.String>. A `StringBuilder` can be dereferenced and modified by the callee, provided it does not exceed the capacity of the `StringBuilder`. It can also be initialized to a fixed length. For example, if you initialize a `StringBuilder` buffer to a capacity of `N`, the marshaler provides a buffer of size (`N`+1) characters. The +1 accounts for the fact that the unmanaged string has a null terminator while `StringBuilder` does not.
+
+For example, the Windows [`GetWindowText`](/windows/desktop/api/winuser/nf-winuser-getwindowtextw) API function (defined in *Windows.h*) requires that the caller pass a fixed-length character buffer to which the function writes the window's text. `LpString` points to a caller-allocated buffer of size `nMaxCount`. The caller is expected to allocate the buffer and set the `nMaxCount` argument to the size of the allocated buffer. The following example shows the `GetWindowText` function declaration as defined in *Windows.h*.
```

Comment:
`GetWindowText` is defined in *winuser.h*, but one should include *Windows.h* to use it.

https://docs.microsoft.com/windows/desktop/api/winuser/nf-winuser-getwindowtexta#requirements

Reply:
@rpetrusha

Reply:
Opened #12249